### ISSUE TITLE
Questions about writing custom gradient functions in BrainPy

### ISSUE THREAD

**Issue by CloudyDory** [brainpy/BrainPy#637]
Hi, is there a guide on writing custom gradient functions in BrainPy (such as those defined in `brainpy.math.surrogate`)?

**Comment by chaoming0625** [comment]
Thanks for the question. I will add examples and tutorials for you.

**Comment by chaoming0625** [comment]
Currently, customizing a ``surrograte`` gradient function can be implemented in the following ways. 


For example, if we want to replace the Heaviside function $\mathcal{H}(v)$ with the surrogate function of ``sigmoid``, we can code with:



```python

import brainpy.math as bm
import jax.numpy as jnp
from jax.scipy.special import expit

class MySurrogate(bm.Surrogate):
  def __init__(self, alpha=1.):
     super().__init__()
     self.alpha = alpha

  def surrogate_fun(self, x):
    return sci.special.expit(x)

  def surrogate_grad(self, x):
    sgax = sci.special.expit(x * self.alpha)
    dx = (1. - sgax) * sgax * self.alpha
    return dx

```

where ``surrogate_fun`` define the forward surrogate function $g(x) = \mathrm{sigmoid}(\alpha x) = \frac{1}{1+e^{-\alpha x}}$, and ``surrogate_grad`` define the gradient of the surrogate function $g'(x) = \alpha * (1 - \mathrm{sigmoid} (\alpha x)) \mathrm{sigmoid} (\alpha x)$.

**Comment by chaoming0625** [comment]
Hope this message can help you.

**Comment by chaoming0625** [comment]
It is important to note here the surrogate gradient function defined in brainpy can not only support the **reverse mode autograd** such as ``jax.grad``, ``brainpy.math.grad`` , and ``jax.vjp``, but also support the **forward mode autograd** including ``jax.jvp``, ``brainpy.math.fwdjac`` and others.  Unlike other libraries, this feature is unique in brainpy.

**Comment by CloudyDory** [comment]
Thank you very much for the information! What about using `@jax.custom_gradient` function decorator? I have found some BrainPy examples using this method.

**Comment by chaoming0625** [comment]
Great question. ``@jax.custom_gradient`` can only support reverse mode autograd. For online learning methods we are going to release, forward mode autograd is important when supporting the surrogate gradient of spikes.

**Comment by chaoming0625** [comment]
Actually not, or very few examples.  However for biological plausibility and scalability, forward grad is much better. 

For a matmul $x W = y$, the forward grad does not require the transpose of $W$ when computing $\partial y / \partial x$. however, reverse grad needs. Therefore, for distributed computing and biological plausibility (the brain does not implement the weight transpose), forward gradients are what we need.
